### ISSUE TITLE
Add Ability to Assign Same Quiz to More than one Course?

### ISSUE THREAD

**Issue by dpentony** [MyCaert/Issues-Enhancements#162]
When creating a quiz or editing a quiz's parameters, you have the ability to select the course the quiz is to be assigned to. See screenshot below. Currently, you can only select one course. You can't use the Ctrl key to select more than one course from the list in the pull down. Could the ability to do so be added? If so, what is the estimate? We need to think this through fully to insure that adding this functionality would not screw how the scores are calculated in the respective courses.

![image](https://cloud.githubusercontent.com/assets/5790325/9945740/aea0e24e-5d54-11e5-9108-9f0bb42a6150.png)

**Comment by bguerrettaz** [comment]
So this one is a little tricky.  When creating a quiz, we can change the
code to accept multiple quiz classes and do database inserts for each quiz
class using the same parameters for each.

Estimate for that is probably 5-7 hours.

The tricky part is editing.  When editing a quiz, you are changing an
entity, if you were to change the quiz class, it just updates the record to
that quiz class ID, but when you are adding more than one to a quiz that's
already built, you'd have to separate the existing quiz out, update that
with any parameters that have changed, and then do inserts for the other
quiz classes that you've added.  This isn't impossible, just some more
work.  Probably looking at 8-12 hours for that.

In my opinion, I think this makes 'more' sense from the quiz creation step,
than the edit step, but I can see why both are wanted.  Both would be a
nice thing to have for the teachers.  Just my 2 cents.

Thanks and let me know how to proceed.

Blake

**Comment by dpentony** [comment]
Hmmm... gonna have to think about this. I'm not sure it is worth doing without being able to do it in the edit process.

Question - could it be done in the edit process but not in the quiz creation process?

**Comment by bguerrettaz** [comment]
Yes, the two changes are agnostic from each other, so you'd be fine to do
it in the edit and not the create, or vice versa.

Blake

**Comment by dpentony** [comment]
We want to move forward with the ability to do this only in the Edit process. That is where it is most needed.

**Comment by bguerrettaz** [comment]
Ok, I have finally completed the changes, tested considerably and made them live as of 15 minutes ago.

Please test as throughly as possible to assure there are no issues with the implementation.

I've taken complete backups of the code files, so if we need to back the changes out, I can do so within minutes.

Please test and confirm changes and as desired.

Thanks,
Blake

**Comment by dpentony** [comment]
Thanks. We'll share with our customers and do some additional testing.

**Comment by dpentony** [comment]
We've discovered an issue when you assign a quiz to a second course. A student is unable to open the test in the second course. See under Mindy's IL CTE account. The name of the test is Business Essay Quiz. She created the quiz for her Entrepreneurship course. Student Brian Ash can access the quiz under that course. She then edited the quiz and assigned it to two other courses. When Brian Ash tries to open the quiz under the other courses he sees the following:

![image](https://cloud.githubusercontent.com/assets/5790325/10233995/71273d7e-6857-11e5-807a-fc5c11903873.png)

When he clicks to open the quiz itself he get this error message:

![image](https://cloud.githubusercontent.com/assets/5790325/10234011/921fd9be-6857-11e5-9335-2b0c063bb006.png)

**Comment by dpentony** [comment]
Here is the message we received from a teacher who was testing this. It is the same problem as we encountered above:

"I thought the upgrade worked. It allowed me to assign a quiz to multiple classes but I did not realize until students brought it to my attention, the quizzes all go into a category requiring them to be administered by an administrator- so they get an error message indicating there are no questions for the selected quiz. The students were unable to access the quizzes from home. I went back to the old way and made individual quizzes for each group. I create the quiz and add a different letter (a, b, or c) after the title and then just add questions from the a version so they all have the same questions. Please let me know if I did something wrong or if it a kink you need to work out."

**Comment by bguerrettaz** [comment]
Ok, I found two issues, which I'm glad we found with testing.

Here is a breakdown of the two issues:

-On the copy, there were three fields that were set that didn't need to be,
this ix fixed.
-For test questions not being there, I needed to insert the questions from
the test being copied into the new test copy.  I completely forgot that we
would want the test questions associated with the test to go with all of
the copies (not sure how I forgot that).  I have created the SQL/Code for
this and have now implemented.

Please test and confirm.  Everything should be good to go now.

Blake

**Comment by dpentony** [comment]
Is there any way to seen when a test was assigned to a second course? Jay Solomonson is an IL ag teacher. He has an test called Exam - Surveying. It looks like he copied it to a second course but the questions did not copy over. When students go to take the test for the second course it was assigned to, they get the same error message as before. See below. Was this test copied over to the second course before you fixed the issue or is it still present?

The teacher is not there today and we are trying to help the sub.

![image](https://cloud.githubusercontent.com/assets/5790325/10369477/14da5ca6-6da0-11e5-829d-2e41d32ff326.png)

**Comment by bguerrettaz** [comment]
It might have been copied before my fix was in place.

Could we try to copy it again from the source quiz to make sure its working
as desired?  That should take care of it.  If not, I'll have to look into
why this one didn't work.

Blake

**Comment by dpentony** [comment]
I did some further testing under my account and did not have any issues. All of the questions showed up and students could access the test in both courses.

@mnichols13 - perhaps we should send Jay Solomonson an email explaining that the first time he copied that test to a second course, it did not copy correctly. We need to see if he would be willing to delete the test in the second course and then re-copy it to make sure all is well. I'd hate to try that without him giving the OK.

**Comment by dpentony** [comment]
We noticed that when you assign a test to a second course and then go to change the activate/de-activate dates for the test under quiz parameters, the changed dates do not stick. See below. This is on David Swanson's (MN) Arc Welding Safety tests that is assigned to two different courses. When he goes to change the highlighted dates, it says you are successful, but the new dates do not stick.

![image](https://cloud.githubusercontent.com/assets/5790325/10466666/11cd0986-71bb-11e5-86f5-beb0118970ee.png)

**Comment by dpentony** [comment]
This seems to be the case in the instance when the quiz was assigned to no course.

**Comment by bguerrettaz** [comment]
This is the only time this happens?  Is this a scenario we want to allow
the user to do?

I'll wait to hear back from you before doing analysis on this issue.

Blake

**Comment by dpentony** [comment]
It seemed to be only on this quiz when assigned to a second course. We could not get the revised  dates to stick on the parameters page when we changed them.

**Comment by dpentony** [comment]
Below is an email exchange Mindy just had with a teacher on this issue:

Steven,
When you pull up the parameters page and it asks you to choose the course, click on NO QUIZ CLASS, then it seems to work ok.  I changed your Band Saw Quiz with success to today. 

Mindy Nichols 
CAERT, Inc. 
P.O. Box 50 
Danville, IL 61834 

On 10/14/2015 2:38 PM, Steven Ellis wrote: 
New Web Contact Request!

You have revieved a new Contact Request.
Name: Steven Ellis
Email: sellis@centralcougars.org
State: Illinois
School: Central Community High School #71
Profile: CTE
Question:
I cant get my band saw or other safety tests to update parameters. (ex. change the date from september to oct. It will allow me the change click save and won't update the date or allow students to view it.

**Comment by dpentony** [comment]
@bguerrettaz - this seems to have to do with the fact that you must re-pick a course to assign the test to when you change the activate/de-activate dates. If you just change the dates on the parameters page and click save, it says that the changes were successful, but the new dates do not stick. The only way to get the dates to stick is to change the dates AND rechoose the course the test is assigned to via the pull down menu. It forces you to re-choose the course the test is assigned to before the changed dates will stick. I am thinking this has to do with the new drop down menu of courses that was developed as part of adding the ability to assign the quiz to more than one course.

I also just noticed that you cannot change ANY of the parameters and get them to stick unless you re-choose a course to assign the test to. This is not how the page worked before. You could change any parameter, save it, and the change would stick. You did not have to re-choose the course the test is assigned to.

I also just noticed that this problem only exists on quizzes that are assigned to NO Quiz Class (None) as the course. Look under my IL ag profile. In active quizzes, I have three instances of the ABC quiz. The problems described above only relate to the instance where it is assigned to no quiz class. The parameter changes will not stick until you re-choose the course from the pull down. However, if you change the parameters on the same quiz that is assigned to the 8.11.14 course, the changes will stick without re-choosing a course. 

In a nutshell if the test had previously been assigned to NO Quiz class and you change the parameters, the course (None) does not stick even though that is not the parameter that has been changed.

**Comment by bguerrettaz** [comment]
Ok, I'm looking into this and will let you know.

One question, is it applicable to assign a quiz to no quiz class? Meaning, should that option be available?

Blake

**Comment by dpentony** [comment]
Yes, it is applicable and should be available. We've always had the ability to assign a quiz to no quiz class. This simply allows students in all courses to see that quiz.

**Comment by bguerrettaz** [comment]
Ok, I have this issue resolved.  No quiz class deemed to be an issue when
selecting multiple quizzes because no quiz class is actually blank or empty.

I've put code in as a way to catch the no quiz class scenario.

Please test and confirm.

Thanks,
Blake

**Comment by dpentony** [comment]
Looks good. Thanks

**Comment by bguerrettaz** [comment]
Is there an issue with this one?  I noticed it was reopened.

**Comment by dpentony** [comment]
I believe there is still an issue when assigning a quiz to a second course.

When you assign a test to no quiz class (which is one of the choices in the pull down on the parameters page), it shows on the Edit Quiz page that the test is assigned to "None". That is correct. See below:

![image](https://cloud.githubusercontent.com/assets/5790325/11092978/fd786748-884c-11e5-83c3-32d257aaa1cf.png)

But if you go to the quiz parameters page for that same quiz, No Quiz Class is not highlighted in the pull down menu. This occurs only when No Quiz Class is chosen. If you choose any other quiz class, it is highlighted in the pull down menu. See below

![image](https://cloud.githubusercontent.com/assets/5790325/11093071/8fe675a2-884d-11e5-9730-b9bdb5083a18.png)

We need No Quiz Class to be highlighted in the pull down menu on the parameters page when a quiz is not assigned to a any course.

**Comment by bguerrettaz** [comment]
This is now fixed.  I have accounted for none/No Quiz Class in the code and made the change.

Please test and confirm.

Thanks,
Blake

**Comment by dpentony** [comment]
Looks good. Thanks.
